WalletWasabi/Tor/TorMonitor.cs

```diff
@@ -13,120 +14,75 @@ namespace WalletWasabi.Tor
-									{
-										if (torEx.RepField == RepField.HostUnreachable)
-										{
-											Uri baseUri = new Uri($"{fallBackTestRequestUri.Scheme}://{fallBackTestRequestUri.DnsSafeHost}");
-											using (var client = new TorHttpClient(baseUri, torSocks5EndPoint))
-											{
-												var message = new HttpRequestMessage(HttpMethod.Get, fallBackTestRequestUri);
-												await client.SendAsync(message, MonitorCts.Token).ConfigureAwait(false);
-											}
+									var message = new HttpRequestMessage(HttpMethod.Get, FallBackTestRequestUri);
+									await client.SendAsync(message, token).ConfigureAwait(false);
+								}
 
-											// Check if it changed in the meantime...
-											if (TorHttpClient.LatestTorException is TorSocks5FailureResponseException torEx2 && torEx2.RepField == RepField.HostUnreachable)
-											{
-												// Fallback here...
-												RequestFallbackAddressUsage = true;
-											}
-										}
-									}
-									else
-									{
-										Logger.LogInfo($"Tor did not work properly for {(int)torMisbehavedFor.TotalSeconds} seconds. Maybe it crashed. Attempting to start it...");
-										await TorProcessManager.StartAsync(ensureRunning: true).ConfigureAwait(false); // Try starting Tor, if it does not work it'll be another issue.
-									}
+								// Check if it changed in the meantime...
+								if (TorHttpClient.LatestTorException is TorSocks5FailureResponseException torEx2 && torEx2.RepField == RepField.HostUnreachable)
+								{
+									// Fallback here...
+									RequestFallbackAddressUsage = true;
 								}
 							}
 						}
-						catch (Exception ex) when (ex is OperationCanceledException || ex is TaskCanceledException || ex is TimeoutException)
-						{
-							Logger.LogTrace(ex);
-						}
-						catch (Exception ex)
+						else
 						{
-							Logger.LogDebug(ex);
+							Logger.LogInfo($"Tor did not work properly for {(int)torMisbehavedFor.TotalSeconds} seconds. Maybe it crashed. Attempting to start it...");
+							bool started = await TorProcessManager.StartAsync(ensureRunning: true).ConfigureAwait(false); // Try starting Tor, if it does not work it'll be another issue.
+
+							if (started)
+							{
+								TorHttpClient.TorDoesntWorkSince = null;
+							}
 						}
 					}
 				}
-				finally
-				{
-					Interlocked.CompareExchange(ref _monitorState, StateStopped, StateStopping); // If IsStopping, make it stopped.
-				}
-			});
-		}
-
-		public async Task StopAsync()
-		{
-			Interlocked.CompareExchange(ref _monitorState, StateStopping, StateRunning); // If running, make it stopping.
-
-			MonitorCts?.Cancel();
-			while (Interlocked.CompareExchange(ref _monitorState, StateStopped, StateNotStarted) == StateStopping)
+			}
+			catch (Exception ex) when (ex is OperationCanceledException || ex is TimeoutException)
+			{
+				Logger.LogTrace(ex);
+			}
+			catch (Exception ex)
```

Comment:
Periodicrunner can handle exceptions, you don't need to catch. Check if OperationCanceledException is handled too in PeriodicRunner.

Reply:
Right.

* PeriodicRunner updated for OperationCanceledException (parent of the original exception) here: https://github.com/zkSNACKs/WalletWasabi/pull/4648/commits/1fa09b03fa327eef9fa02c74f3b5e2014ed79781
* TorMonitor is then updated here: https://github.com/zkSNACKs/WalletWasabi/pull/4648/commits/13b5e6a532fbbc065314721302ee87c93fafbd9f